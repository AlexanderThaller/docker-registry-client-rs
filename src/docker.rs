use std::{
    collections::HashMap,
    sync::Arc,
};

use reqwest::{
    header::HeaderMap,
    Client as HTTPClient,
};
use tokio::sync::RwLock;
use url::Url;

use crate::{
    image_name::{
        ImageName,
        Registry,
    },
    manifest::Manifest,
};

mod error;

pub use error::Error;

#[derive(Debug, Default, Clone)]
pub struct Client {
    client: HTTPClient,

    token_cache: Arc<RwLock<HashMap<CacheKey, String>>>,
}

#[derive(Debug, Eq, PartialEq, Hash)]
struct CacheKey {
    registry: Registry,
    repository: String,
    image_name: String,
}

#[derive(Debug, Clone)]
pub struct Response {
    pub digest: Option<String>,
    pub manifest: Manifest,
}

impl Client {
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    /// # Errors
    /// Returns an error if the request fails.
    /// Returns an error if the response body is not valid JSON.
    /// Returns an error if the response body is not a valid manifest.
    /// Returns an error if the response status is not successful.
    pub async fn get_manifest(&self, image_name: &ImageName) -> Result<Response, Error> {
        let mut headers = self.get_headers(image_name).await?;

        let accept_header = [
            "application/vnd.docker.container.image.v1+json",
            "application/vnd.docker.distribution.manifest.list.v2+json",
            "application/vnd.docker.distribution.manifest.v2+json",
            "application/vnd.docker.image.rootfs.diff.tar.gzip",
            "application/vnd.docker.image.rootfs.foreign.diff.tar.gzip",
            "application/vnd.docker.plugin.v1+json",
            "application/vnd.oci.image.index.v1+json",
            "application/vnd.oci.image.manifest.v1+json",
        ]
        .join(", ");

        headers.insert(
            "Accept",
            accept_header
                .parse()
                .map_err(Error::ParseManifestAcceptHeader)?,
        );

        let registry_domain = image_name.registry.registry_domain();

        let url = Url::parse(&format!(
            "https://{}/v2/{}/{}/manifests/{}",
            registry_domain, image_name.repository, image_name.image_name, image_name.identifier
        ))
        .map_err(Error::InvalidManifestUrl)?;

        let response = self
            .client
            .get(url.as_str())
            .headers(headers)
            .send()
            .await
            .map_err(Error::GetManifest)?;

        let status = response.status();
        let digest = response
            .headers()
            .get("Docker-Content-Digest")
            .map(|header| {
                header
                    .to_str()
                    .map(String::from)
                    .map_err(Error::ParseDockerContentDigestHeader)
            })
            .transpose()?;

        let body = response.text().await.map_err(Error::ExtractManifestBody)?;

        if !status.is_success() {
            if status == reqwest::StatusCode::NOT_FOUND {
                return Err(Error::ManifestNotFound(url));
            }

            return Err(Error::FailedManifestRequest(status, body));
        }

        let manifest =
            serde_json::from_str(&body).map_err(|e| Error::DeserializeManifestBody(e, body))?;

        Ok(Response { digest, manifest })
    }

    async fn get_headers(&self, image_name: &ImageName) -> Result<HeaderMap, Error> {
        #[derive(Debug, serde::Deserialize)]
        struct Token {
            token: String,
        }

        let cache_key = CacheKey {
            registry: image_name.registry.clone(),
            repository: image_name.repository.clone(),
            image_name: image_name.image_name.clone(),
        };

        let token_cache = self.token_cache.read().await;
        let token = token_cache.get(&cache_key).cloned();
        drop(token_cache);

        let token = if let Some(token) = token {
            token
        } else {
            let token_url = match image_name.registry {
                Registry::Github => format!(
                    "https://ghcr.io/token?scope=repository:{}/{}:pull&service=ghcr.io",
                    image_name.repository, image_name.image_name
                ),
                Registry::DockerHub => format!("https://auth.docker.io/token?service=registry.docker.io&scope=repository:{}/{}:pull&service=registry.docker.io", image_name.repository, image_name.image_name),
                Registry::Quay => format!("https://quay.io/v2/auth?scope=repository:{}/{}:pull&service=quay.io", image_name.repository, image_name.image_name),

                Registry::Specific(_) => return Ok(HeaderMap::new()),
            };

            let token_url = Url::parse(&token_url).map_err(Error::InvalidTokenUrl)?;

            let response = self
                .client
                .get(token_url)
                .send()
                .await
                .map_err(Error::GetToken)?;

            let body = response.text().await.map_err(Error::ExtractTokenBody)?;

            let token: Token =
                serde_json::from_str(&body).map_err(|e| Error::DeserializeToken(e, body))?;

            let mut token_cache = self.token_cache.write().await;
            token_cache.insert(cache_key, token.token.clone());

            token.token
        };

        let mut headers = HeaderMap::new();
        headers.insert(
            "Authorization",
            format!("Bearer {token}")
                .parse()
                .map_err(Error::ParseAuthorizationHeader)?,
        );

        Ok(headers)
    }
}
