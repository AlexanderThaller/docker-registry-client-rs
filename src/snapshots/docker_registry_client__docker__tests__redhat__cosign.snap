---
source: src/docker.rs
expression: response
---
{
  "digest": "sha256:9d50ceb15f023eda8f58032849eedc0216236d2e2f4cfe1cdf97c00ae7798cfe",
  "manifest": {
    "schemaVersion": 2,
    "mediaType": "application/vnd.oci.image.index.v1+json",
    "manifests": [
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 2474,
        "digest": "sha256:bb6fdb21aff84bcbf2b4d9cd179a4f5ecdad7cf4ae8544fdd8272f3a23a43093",
        "platform": {
          "architecture": "amd64",
          "os": "linux"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 2474,
        "digest": "sha256:6bdc038ccf293582a753ca587bd0a8cb87f0db27368ed6249f8b28ec5fb112e2",
        "platform": {
          "architecture": "arm64",
          "os": "linux",
          "variant": "v8"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 2474,
        "digest": "sha256:236c55878684b1a235a16f40dfeb1bea5d33cb29ba7589cd5b6e6ea560e63328",
        "platform": {
          "architecture": "arm",
          "os": "linux",
          "variant": "v7"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 2474,
        "digest": "sha256:64d2603e8d8f922450b95831c1b8976e81e23646f586ba27ca6ce53314b66843",
        "platform": {
          "architecture": "s390x",
          "os": "linux"
        }
      },
      {
        "mediaType": "application/vnd.oci.image.manifest.v1+json",
        "size": 2474,
        "digest": "sha256:76fa36442861024a8a006b8d9c8d6b37eea0a9eede205bbc7904a8d3051deee0",
        "platform": {
          "architecture": "ppc64le",
          "os": "linux"
        }
      }
    ]
  }
}
